### PR TITLE
Clippy lints

### LINKED ISSUE #207: Enable all clippy lints
*Note*: migrated from original JIRA: https://issues.apache.org/jira/browse/ARROW-10597

The idea is to get the code to the point where all clippy lints can be enabled. 

Here is a list of clippy lints that was disabled:

https://github.com/apache/arrow/pull/8666/files#diff-678f16a0c102ed15656bd00d98c47755920c8c4ede296e3bf5c09a0d4f38a42cR19

The goal of this ticket is to enable them (maybe it would be better to break it into multiple PRs/subtasks)

### PR DESCRIPTION
**Pull request by LouisGariepy**
# Which issue does this PR close?

Closes #207.

# Rationale for this change

This PR fixes most of the remaining clippy `allow`s remaining in the code base.

Clippy `allow`s sometimes silence actually useful lints. Even if the lint was not useful at some point, it's easy to forget about them even after the code has been fixed.

For these reasons it's good practice to clean them up once in a while.

# What changes are included in this PR?

The bulk of the changes is simply removing `allows` the were not needed at all.

Some changes are slightly more involved, but should still be pretty localized and self-explanatory.

# Are these changes tested?

This PR does not (and should not) introduce new features or modify existing features. Thus, existing test coverage should be sufficient.

I added one test to for the code I added for data pointer equality (related to the removal of [vtable address comparison](https://rust-lang.github.io/rust-clippy/master/index.html#vtable_address_comparisons)  lints

# Are there any user-facing changes?

No(*) (see the Notes section below)

# Notes
All of the committed fixes are fairly straightforward and uncontroversial.

Some fixes would be breaking(*) or require maintainer input to get right. I documented these cases with `TODO(clippy)` and gave ample details and potential solutions. *I would appreciate if the reviewer could pay special attention to these so we can decide which ones should be fixed as part of this PR, which ones should be fixed as part of a separate PR, and which ones should not be fixed at all*.

The remaining `allow`s are all [too many arguments](https://rust-lang.github.io/rust-clippy/master/index.html#too_many_arguments) lints whose fixes are not straightforward and might require some more in-depth redesign. I'm proposing to leave them as is for now and fix them on a case-by-case basis as solutions become more evident. I did not mark these with `TODO(clippy)` as I can't suggest solutions for them without some more careful examination. I will try to get to them in due time, in one or more future PRs.

### CODE REVIEW

**Comment by LouisGariepy** [comment]
Oh one thing I was unsure about:

I put the `DataPtr` trait in `datafusion_common::utils`. I'm not sure that's the right place for this utility. Please let me know :smiley: .

**Comment by LouisGariepy** [comment]
Also, bear in mind that the line count includes several added detailed comments. I'm fairly certain that overall this PR removes code or is very close to neutral.

**Comment by alamb** [comment]
Thank you @LouisGariepy  -- I plan to review this carefully tomorrow or Tuesday if no one beats me to it

File: datafusion/core/src/avro_to_arrow/reader.rs [hunk 1]
```diff
@@ -153,7 +153,15 @@ impl<'a, R: Read> Reader<'a, R> {
     }
 
     /// Returns the next batch of results (defined by `self.batch_size`), or `None` if there
-    /// are no more results
+    /// are no more results.
+    //
+    // TODO(clippy): The clippy `allow` could be removed by renaming this method to `next_batch`.
+    // This would also make the intent of the method clearer.
+    //
+    // Another option could be to rework `AvroArrowArrayReader::next_batch` so it returns an
```
Comment (alamb) [inline datafusion/core/src/avro_to_arrow/reader.rs:161]: I think this second option would be my preference

File: datafusion/core/src/catalog/information_schema.rs [hunk 2]
```diff
@@ -486,37 +486,35 @@ struct InformationSchemaColumnsBuilder {
...
 
-        self.ordinal_positions.append_value(column_position as u64);
+        self.ordinal_positions.append_value(field_position as u64);
 
         // DataFusion does not support column default values, so null
         self.column_defaults.append_null();
 
         // "YES if the column is possibly nullable, NO if it is known not nullable. "
-        let nullable_str = if is_nullable { "YES" } else { "NO" };
+        let nullable_str = if field.is_nullable() { "YES" } else { "NO" };
```
Comment (alamb) [inline datafusion/core/src/catalog/information_schema.rs:512]: this is a nice cleanup

File: datafusion/core/src/catalog/mod.rs [hunk 3]
```diff
@@ -18,6 +18,10 @@
 //! This module contains interfaces and default implementations
 //! of table namespacing concepts, including catalogs and schemas.
 
+// TODO(clippy): Having a `datasource::datasource` module path is unclear and ambiguous.
```
Comment (alamb) [inline datafusion/core/src/catalog/mod.rs:21]: 👍

File: datafusion/core/src/physical_optimizer/pipeline_fixer.rs [hunk 4]
```diff
@@ -677,8 +677,8 @@ mod hash_join_tests {
             ..
         }) = plan.as_any().downcast_ref::<HashJoinExec>()
         {
-            let left_changed = Arc::ptr_eq(left, &right_exec);
-            let right_changed = Arc::ptr_eq(right, &left_exec);
+            let left_changed = Arc::data_ptr_eq(left, &right_exec);
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/pipeline_fixer.rs:680]: I like the fact that you have moved the (potentially dubious) pointer comparisons into a trait that has much better documentation about rationale. 🙏  thank you

File: datafusion/core/src/physical_plan/aggregates/mod.rs [hunk 5]
```diff
@@ -551,52 +549,21 @@ impl AggregateExec {
...
             Ok(StreamType::AggregateStream(AggregateStream::new(
-                self.mode,
-                self.schema.clone(),
-                self.aggr_expr.clone(),
-                self.filter_expr.clone(),
-                input,
-                baseline_metrics,
-                context,
-                partition,
+                self, context, partition,
```
Comment (alamb) [inline datafusion/core/src/physical_plan/aggregates/mod.rs:554]: this certainly looks much nicer 😍

File: datafusion/core/src/physical_plan/file_format/parquet/row_groups.rs [hunk 6]
```diff
@@ -263,6 +263,15 @@ mod tests {
     };
     use std::sync::Arc;
 
+    struct PrimitiveTypeField<'a> {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/file_format/parquet/row_groups.rs:266]: 👍

File: datafusion/core/src/physical_plan/file_format/parquet/row_groups.rs [hunk 7]
```diff
@@ -263,6 +263,15 @@ mod tests {
...
+        name: &'a str,
```
Comment (alamb) [inline datafusion/core/src/physical_plan/file_format/parquet/row_groups.rs:267]: I don't think it is important but given that `name` seems to always be a constant string, you could do something like

```suggestion
        name: &'static str,
```

And avoid having to have the named lifetime `'a` on this struct

File: datafusion/core/src/physical_plan/file_format/parquet/row_groups.rs [hunk 8]
```diff
@@ -272,14 +281,14 @@ mod tests {
...
-        let schema_descr = get_test_schema_descr(vec![(
-            "c1",
-            PhysicalType::INT32,
-            None,
-            None,
-            None,
-            None,
-        )]);
+        let schema_descr = get_test_schema_descr(vec![PrimitiveTypeField {
+            name: "c1",
```
Comment (alamb) [inline datafusion/core/src/physical_plan/file_format/parquet/row_groups.rs:285]: I am a big fan of explicit field names as a way to better document what is going on here.

File: datafusion/core/src/physical_plan/file_format/parquet/row_groups.rs [hunk 9]
```diff
@@ -347,8 +356,22 @@ mod tests {
...
-            ("c1", PhysicalType::INT32, None, None, None, None),
-            ("c2", PhysicalType::INT32, None, None, None, None),
+            PrimitiveTypeField {
+                name: "c1",
+                physical_ty: PhysicalType::INT32,
+                logical_ty: None,
+                precision: None,
+                scale: None,
+                byte_len: None,
+            },
```
Comment (alamb) [inline datafusion/core/src/physical_plan/file_format/parquet/row_groups.rs:366]: When I was reading this, I was thinking these tests might be even more readable if they had a builder style, something like

```rust
 PrimtiveTypeField::new("c1", PhysicalType::INT32)
```

And then for tests that need other optional fields set:

```
 PrimtiveTypeField::new("c1", PhysicalType::INT32)
  .with_logical_type(LogicalType::Decimal {
                scale: 0,
                precision: 9,
            });
```

File: datafusion/expr/src/logical_plan/extension.rs [hunk 10]
```diff
@@ -84,6 +84,12 @@ pub trait UserDefinedLogicalNode: fmt::Debug + Send + Sync {
     /// of self.inputs and self.exprs.
     ///
     /// So, `self.from_template(exprs, ..).expressions() == exprs
+    //
+    // TODO(clippy): This should probably be renamed to use a `with_*` prefix. Something
+    // like `with_template`, or `with_exprs_and_inputs`.
+    //
+    // Also, I think `ExtensionPlanNode` has been renamed to `UserDefinedLogicalNode`
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/extension.rs:91]: 👍

**Review by alamb — APPROVED** [review]
Thank you very much @LouisGariepy  - I went through this PR carefully and I think it is a really nice cleanup and thus plan to merge it in.

I had some comments that might improve the code more, but I think this PR as it is is a great step forward and we can address the other items as follow on PRs.

Again, thank you for helping keep the codebase cleaner 🦾

**Comment by LouisGariepy** [comment]
@alamb Thank you for the review :smiley:. I implemented your suggestions and fixed a typo that slipped in.

Just to be sure, regarding the `TODO(clippy)` comments, would you prefer if those would be fixed in a separate PR? I don't mind fixing them in this PR with your answers, but I don't want to make this PR too scattered.

**Comment by alamb** [comment]
> Just to be sure, regarding the TODO(clippy) comments, would you prefer those to be fixed in this PR, or in a separate one?

I think a separate one is better -- I don't see any reason to hold up the changes in this PR for technically unrelated other changes.

THanks again @LouisGariepy

File: datafusion/core/src/physical_plan/file_format/parquet/row_groups.rs [hunk 11]
```diff
@@ -562,17 +553,15 @@ mod tests {
         // The decimal of arrow is decimal(5,2), the decimal of parquet is decimal(9,0)
         let schema =
             Schema::new(vec![Field::new("c1", DataType::Decimal128(9, 0), false)]);
-        let schema_descr = get_test_schema_descr(vec![PrimitiveTypeField {
-            name: "c1",
-            physical_ty: PhysicalType::INT32,
-            logical_ty: Some(LogicalType::Decimal {
+
+        let field = PrimitiveTypeField::new("c1", PhysicalType::INT32)
```
Comment (alamb) [inline datafusion/core/src/physical_plan/file_format/parquet/row_groups.rs:557]: ❤️

**Review by alamb — APPROVED** [review]

**Pull request merged**